### ISSUE TITLE
InvalidOperationException thrown while opening a connection

### ISSUE THREAD

**Issue by ddunkin** [mysql-net/MySqlConnector#476]
```
System.InvalidOperationException: The operation is not allowed on non-connected sockets.
   at System.Net.Sockets.TcpClient.GetStream()
   at MySqlConnector.Core.ServerSession.<OpenTcpSocketAsync>d__69.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MySqlConnector.Core.ServerSession.<ConnectAsync>d__56.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MySqlConnector.Core.ConnectionPool.<GetSessionAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at System.Runtime.CompilerServices.ConfiguredValueTaskAwaitable`1.ConfiguredValueTaskAwaiter.GetResult()
   at MySql.Data.MySqlClient.MySqlConnection.<CreateSessionAsync>d__83.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at System.Runtime.CompilerServices.ConfiguredValueTaskAwaitable`1.ConfiguredValueTaskAwaiter.GetResult()
   at MySql.Data.MySqlClient.MySqlConnection.<OpenAsync>d__25.MoveNext()
```

Running MySqlConnector 0.36.1 on .NET Framework 4.7.

This network occasionally has connectivity problems.

**Comment by bgrainger** [comment]
AFAICT, `tcpClient.ConnectAsync` (https://github.com/mysql-net/MySqlConnector/blob/c175893a57fb7d8c502eacb2cc18cea4c19e6aef/src/MySqlConnector/Core/ServerSession.cs#L703) succeeded (no `SocketException` was thrown) but then `tcpClient.Client.Connected` was `false` (so `GetStream` threw). The most likely explanation seems to be that there's a race with the callback passed to `cancellationToken.Register` and it closed the socket after `ConnectAsync` completed successfully. 

We should check `m_socket.Connected` and throw a `MySqlException`.

**Comment by bgrainger** [comment]
Fixed in [0.40.0](https://github.com/mysql-net/MySqlConnector/releases/tag/0.40.0).
